indico/modules/events/registration/api.py

```diff
@@ -7,19 +7,98 @@
+from webargs import fields
 from werkzeug.exceptions import BadRequest, Forbidden
 
 from indico.core import signals
+from indico.modules.events import Event
 from indico.modules.events.controllers.base import RHProtectedEventBase
-from indico.modules.events.models.events import Event
-from indico.modules.events.registration.models.registrations import RegistrationState
+from indico.modules.events.registration.models.forms import RegistrationForm
+from indico.modules.events.registration.models.registrations import Registration, RegistrationState
 from indico.modules.events.registration.util import build_registration_api_data, build_registrations_api_data
-from indico.web.rh import RH, oauth_scope
+from indico.web.args import use_kwargs
+from indico.web.rh import RH, cors, json_errors, oauth_scope
+
+
+@json_errors
+@cors
+@oauth_scope('registrants')
+class RHAPICheckinBase(RHProtectedEventBase):
+    """Base class for the Check-in API."""
+
+
+class RHAPIEvent(RHAPICheckinBase):
+    """Manage a single event."""
+
+    def _process(self):
+        from indico.modules.events.registration.schemas import CheckinEventSchema
+        return jsonify(CheckinEventSchema().dump(self.event))
+
+
+class RHAPIRegFormBase(RHAPICheckinBase):
+    """Base class for registration forms management."""
+
+    def _check_access(self):
+        RHAPICheckinBase._check_access(self)
+        if not self.event.can_manage(session.user, permission='registration'):
+            raise Forbidden()
+
+    def _process_args(self):
+        RHAPICheckinBase._process_args(self)
+        self.regform = (RegistrationForm.query
+                        .filter_by(id=request.view_args['reg_form_id'], is_deleted=False)
+                        .first_or_404())
+
+
+class RHAPIRegForm(RHAPIRegFormBase):
+    """Manage a single registration form."""
+
+    def _process(self):
+        from indico.modules.events.registration.schemas import CheckinRegFormSchema
+        return jsonify(CheckinRegFormSchema().dump(self.regform))
+
+
+class RHAPIRegistrations(RHAPIRegFormBase):
+    """Manage registrations."""
+
+    def _process(self):
+        from indico.modules.events.registration.schemas import CheckinRegistrationSchema
+        registrations = (Registration.query.with_parent(self.regform)
+                         .filter(~Registration.is_deleted)
+                         .all())
+        return jsonify(CheckinRegistrationSchema(many=True).dump(registrations))
+
+
+class RHAPIRegistration(RHAPIRegFormBase):
+    """Manage a single registration."""
+
+    def _process_args(self):
+        RHAPIRegFormBase._process_args(self)
+        registration_id = request.view_args['registration_id']
+        self.registration = (Registration.query
+                             .filter_by(id=registration_id, is_deleted=False)
+                             .join(RegistrationForm, RegistrationForm.id == Registration.registration_form_id)
+                             .filter_by(is_deleted=False)
```

Comment:
Mixing `join` and `filter_by` results in pretty hard-to-understand code since the filter_by is relative to the last join.
Also, with the current code you can get any registration from any event as long as you can manage *some* event (and use that event ID in the URL)...

You can use `.with_parent(self.regform)` to query only registrations from the specified regform.

Using the normalization to make sure all url params are in sync would be a good idea as well:

```python
    normalize_url_spec = {
        'locators': {
            lambda self: self.registration
        }
    }
```

In the old code this is also missing, but `build_registrations_api_data` uses `with_parent(event)` and gets the registration via the regform, so tampering with URL params would not expose anything there.